File: ecs-go/internal/token/token.go
package token

// Kind is the lexical category of a token.
type Kind int

const (
	Unknown    Kind = iota
	OpenTag         // <?php <?= <?
	CloseTag        // ?>
	InlineHTML      // text outside PHP tags
	Whitespace      // spaces, tabs, newlines
	Comment         // // # /* */
	DocComment      // /** */
	Variable        // $foo
	Ident           // names (T_STRING): function/class names, constants, types
	Keyword         // reserved words: function, class, return, namespace, ...
	Number
	String // '...' "..."
	Punct  // operators, braces, ; , etc.
	EOF
)

var kindNames = map[Kind]string{
	Unknown:    "Unknown",
	OpenTag:    "OpenTag",
	CloseTag:   "CloseTag",
	InlineHTML: "InlineHTML",
	Whitespace: "Whitespace",
	Comment:    "Comment",
	DocComment: "DocComment",
	Variable:   "Variable",
	Ident:      "Ident",
	Keyword:    "Keyword",
	Number:     "Number",
	String:     "String",
	Punct:      "Punct",
	EOF:        "EOF",
}

func (k Kind) String() string {
	if n, ok := kindNames[k]; ok {
		return n
	}
	return "Kind(?)"
}

// Token is a single lexical unit. The stream is lossless: concatenating every
// Value in order reproduces the original source byte for byte.
type Token struct {
	Kind  Kind
	Value string
	Pos   int // byte offset in source
}

File: ecs-rust/src/token.rs
// Token kinds and the Token struct, mirroring ecs-go's internal/token.

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Kind {
    OpenTag,    // <?php <?= <?
    CloseTag,   // ?>
    InlineHtml, // text outside PHP tags
    Whitespace, // spaces, tabs, newlines
    Comment,    // // # /* */
    DocComment, // /** */
    Variable,   // $foo
    Ident,      // names (T_STRING): function/class names, constants, types
    Keyword,    // reserved words
    Number,
    String, // '...' "..."
    Punct,  // operators, braces, ; , etc.
}

// A token's bytes are either a span into the original source (the common case,
// zero-allocation) or owned bytes produced when a fixer rewrites the token. This
// copy-on-write model keeps lexing allocation-free while staying byte-lossless.
#[derive(Clone)]
pub enum Value {
    Span(u32, u32), // [start, end) byte offsets into the source
    Owned(Vec<u8>),
}

#[derive(Clone)]
pub struct Token {
    pub kind: Kind,
    pub value: Value,
}

impl Token {
    pub fn span(kind: Kind, start: usize, end: usize) -> Self {
        Token {
            kind,
            value: Value::Span(start as u32, end as u32),
        }
    }

    pub fn owned(kind: Kind, bytes: Vec<u8>) -> Self {
        Token {
            kind,
            value: Value::Owned(bytes),
        }
    }
}
